service.ts

```diff
@@ -0,0 +1,69 @@
+import express from 'express';
+import crypto from 'node:crypto';
+import bodyParser from 'body-parser';
+import { createValidator } from 'express-joi-validation';
+import {
+  createUserBodySchema,
+  paramsIDSchema,
+  paramsSubstringLimitSchema,
+  updateUserBodySchema,
+} from './schemas';
+import type { User } from './types';
+
+const app = express();
+const validator = createValidator();
+
+app.listen(3000);
+
+const users: User[] = [];
+
+app.use(bodyParser.json());
+
+// get all users
+app.get('/users', (req, res) => {
+  res.json(users);
+});
+
+// get user by ID
+app.get('/user/:id', validator.params(paramsIDSchema), (req, res) => {
+  res.json(users.filter((user) => user.id === req.params.id));
+});
+
+// create user
+app.post('/user', validator.body(createUserBodySchema), (req, res) => {
+  const user = { ...req.body, id: crypto.randomUUID() };
+  users.push(user);
+  res.json(user);
+});
+
+//update user
+app.put(
+  '/user/:id',
+  validator.params(paramsIDSchema),
+  validator.body(updateUserBodySchema),
+  (req, res) => {
+    const index = users.findIndex((user) => user.id === req.params.id);
+    users[index] = { ...users[index], ...req.body };
+    res.json(users[index]);
+  }
+);
+
+//delete user
+app.put('/user/delete/:id', validator.params(paramsIDSchema), (req, res) => {
+  const index = users.findIndex((user) => user.id === req.params.id);
+  users[index] = { ...users[index], isDeleted: true };
+  res.json(users[index]);
+});
+
+// filter users on limit
+app.get(
+  '/users/:loginSubstring/:limit',
```

Comment:
I'm afraid I misguided you when you asked about this task's checkpoint. I'm really-really sorry.
Actually this would be endpoint `/users`, but with query params. So, if there are query params, then you return limited users filtered by substring, otherwise you return all users.
I will talk about this tomorrow on Q&A.

Reply:
fixed